File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.js
/**
 * @param {number[]} difficulty
 * @param {number[]} profit
 * @param {number[]} worker
 * @return {number}
 */
var maxProfitAssignment = function(difficulty, profit, worker) {
    /*
        4,5,6,7 -> 
        4 -> 20
        5 -> 20
        6 -> 30
        7 -> 30
    */
    let maxProfits = 0

    let wi = 0 
    let jobs = []
    for ( let i = 0; i<difficulty.length; i++) {
        jobs.push([difficulty[i], profit[i]])
    }
    jobs.sort((a, b) => a[0] -b[0])
    worker.sort((a, b) => a - b)
    let ans = 0
    // difficulty.sort((a, b) => a - b)
    // worker.sort((a,b) => a - b)
    for ( let i = 0; i<worker.length; i++) {
        while(wi < jobs.length && jobs[wi][0] <= worker[i]) {
            maxProfits = Math.max(maxProfits, jobs[wi][1])
            wi += 1
        }
        ans += maxProfits
    }

    return ans
};

File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.go
func maxProfitAssignment(difficulty []int, profit []int, worker []int) int {
    n := len(difficulty)

    jobs := make([][2]int, n)

    for i := 0; i< n; i++ {
        jobs[i] = [2]int{difficulty[i], profit[i]}
    }

    sort.Slice(jobs, func(i, j int) bool {
        return jobs[i][0] < jobs[j][0]
    })

    sort.Ints(worker)

    ans := 0
    maxProfits := 0
    wi := 0

    for _, w := range worker {
        for ; wi < n && jobs[wi][0] <= w; wi++ {
            if jobs[wi][1] > maxProfits {
                maxProfits = jobs[wi][1]
            }
        }
        ans += maxProfits
    }

    return ans
}